finalitygadget/finalitygadget.go

```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
+				if height > math.MaxInt64 {
+					return fmt.Errorf("block height %d exceeds maximum int64 value", height)
+				}
+				wg.Add(1)
+				go func(h uint64) {
+					defer wg.Done()
+					results <- fg.processHeight(h)
+				}(height)
 			}
 
-			// Check the block is babylon finalized using sdk client
-			isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
-			if err != nil && !errors.Is(err, types.ErrBtcStakingNotActivated) {
-				return fmt.Errorf("error checking block %d: %v", block.BlockHeight, err)
+			// Close results channel once all goroutines complete
+			go func() {
+				wg.Wait()
+				close(results)
+			}()
+
+			// Process results and extract error (if any)
+			statusMap := make(map[uint64]blockResult)
+			var processingError error
+			for result := range results {
+				if result.err != nil {
+					processingError = result.err
+					break
+				}
+				statusMap[result.height] = result
 			}
-			// If not finalized, throw error
-			if !isFinal {
-				return fmt.Errorf("block %d should be finalized according to client but is not", block.BlockHeight)
+			if processingError != nil {
+				return processingError
 			}
 
-			// If finalised, store the block in DB
-			err = fg.InsertBlock(block)
-			if err != nil {
-				return fmt.Errorf("error storing block %d: %v", block.BlockHeight, err)
+			// Find the last consecutive finalized block
+			var lastFinalizedHeight uint64
+			for height := startHeight; height <= endHeight; height++ {
+				status, exists := statusMap[height]
+				if !exists || !status.isFinalized {
+					break
+				}
+				lastFinalizedHeight = height
+			}
+
+			// If no blocks were finalized, wait for next poll
+			if lastFinalizedHeight < startHeight {
+				return nil
+			}
+
+			// Batch insert all consecutive finalized blocks
+			finalizedBlocks := make([]*types.Block, lastFinalizedHeight-startHeight+1)
+			for height := startHeight; height <= lastFinalizedHeight; height++ {
+				finalizedBlocks[height-startHeight] = statusMap[height].block
+			}
+			if err := fg.insertBlocks(finalizedBlocks); err != nil {
+				return fmt.Errorf("error storing blocks: %w", err)
 			}
-			fg.lastProcessedHeight = block.BlockHeight
-			fg.logger.Info("Inserted new finalized block", zap.Uint64("block_height", block.BlockHeight))
+			fg.lastProcessedHeight = lastFinalizedHeight
+
+			// Update start height for next batch
+			startHeight = lastFinalizedHeight + 1
 		}
 	}
 
 	return nil
 }
 
+func (fg *FinalityGadget) processHeight(height uint64) blockResult {
+	result := blockResult{height: height}
+
+	// Fetch block from rpc
+	block, err := fg.queryBlockByHeight(int64(height))
+	if err != nil {
+		result.err = fmt.Errorf("error getting block at height %d: %w", height, err)
+		return result
+	}
+	result.block = block
+
+	// Check finalization
+	isFinalized, err := fg.QueryIsBlockBabylonFinalizedFromBabylon(block)
+	if err != nil && !errors.Is(err, types.ErrBtcStakingNotActivated) {
```

Comment:
`processBlocksTillHeight` should never be called when btc staking is not activated

so having this check if unnecessary and can cause confusion (i.e. what should the behavior of this function be when it's not btc staking activated)

Reply:
now addressed by latest commit